include/fi_mem.h

```diff
@@ -69,24 +69,31 @@ static inline void *mem_dup(const void *src, size_t size)
  */
 #define FREESTACK_EMPTY	NULL
 
+#define FREESTACK_HEADER 					\
+	size_t		size;					\
+	void		*next;					\
+
 #define freestack_isempty(fs)	((fs)->next == FREESTACK_EMPTY)
 #define freestack_push(fs, p)					\
 {								\
 	*(void **) p = (fs)->next;				\
 	(fs)->next = p;						\
 }
-#define freestack_pop(fs)					\
-	({							\
-		void *p = (fs)->next;				\
-		assert(!freestack_isempty(fs));			\
-		(fs)->next = *((void **) p);			\
-		p;						\
-	})
+#define freestack_pop(fs) freestack_pop_impl(fs, fs->next)
+
+static inline void* freestack_pop_impl(void *fs, void *fs_next)
+{
+	struct {
+		FREESTACK_HEADER;
+	} *freestack = fs;
+	assert(!freestack_isempty(freestack));
+	(freestack)->next = *((void **)fs_next);
```

Comment:
The parentheses around this "freestack" is not needed now that it is not a macro anymore.